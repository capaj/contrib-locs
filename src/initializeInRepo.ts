import git, { Commit } from 'nodegit'
import path from 'path'
import micromatch from 'micromatch'
import { LocsStatsPerUser } from './LocsStats'
import { getConfig } from './getConfig'
import execa from 'execa'
import debug from 'debug'

const log = debug('contrib-locs')

export const initializeInRepo = async (repoRootPath: string) => {
  const statInstance = new LocsStatsPerUser(false)
  const config = await getConfig()

  const defaultBranchName = execa.sync(
    'git',
    ['symbolic-ref', '--short', 'HEAD'],
    { cwd: path.resolve(repoRootPath) }
  ).stdout
  log('using default branch', defaultBranchName)

  const repo = await git.Repository.open(
    path.resolve(path.join(repoRootPath, '.git'))
  )
  const firstCommitOnMaster = await repo.getBranchCommit(defaultBranchName)
  const revwalk = git.Revwalk.create(repo)
  revwalk.reset()
  revwalk.sorting(git.Revwalk.SORT.TIME, git.Revwalk.SORT.REVERSE)
  revwalk.push(firstCommitOnMaster.id())

  // step through all OIDs for the given reference
  const allOids = []
  let hasNext = true
  let commit: Commit | undefined
  while (hasNext) {
    try {
      const oid = await revwalk.next()
      log(`commit: ${oid.tostrS()}`)
      commit = await repo.getCommit(oid)
      const authorEmail = commit.author().email()

      const diff = await commit.getDiff()
      await Promise.all(
        diff.map(async (d) => {
          const patches = await d.patches()

          await Promise.all(
            patches.map(async (patch) => {
              const path = patch.newFile().path()
              if (micromatch.isMatch(path, config.match)) {
                log(`adding a diff for ${path}: `, patch.lineStats())
                statInstance.addLineStat(authorEmail, patch.lineStats())
              }
            })
          )
        })
      )

      allOids.push(oid)
    } catch (err) {
      hasNext = false
      if (!commit) {
        throw err
      } else {
        statInstance.setLastCommit(commit)
      }
    }
  }
  statInstance.countPercentages()
  return statInstance
}
