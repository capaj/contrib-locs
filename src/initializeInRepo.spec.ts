import { initializeInRepo } from './initializeInRepo'

describe('initializeInRepo', () => {
  it('should initialize in solid repo', async () => {
    // const statInstance = await initializeInRepo('.') // TODO use solid repo in fixtures, not current dir, but we need to find a fix for https://github.com/nodegit/nodegit/issues/1679
    const statInstance = await initializeInRepo('fixtures/solid')
    expect(statInstance.output).toMatchSnapshot()
  })
})
