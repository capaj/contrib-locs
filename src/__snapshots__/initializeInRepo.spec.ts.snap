// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initializeInRepo should initialize in solid repo 1`] = `
{
  "1168397+davedbase@users.noreply.github.com": {
    "email": "1168397+davedbase@users.noreply.github.com",
    "loc": 186,
    "percentageOfTotal": 0.022774273886356375,
    "total_additions": 94,
    "total_deletions": 92,
  },
  "aminyahyaabadi74@gmail.com": {
    "email": "aminyahyaabadi74@gmail.com",
    "loc": 122,
    "percentageOfTotal": 0.014937964592126225,
    "total_additions": 121,
    "total_deletions": 1,
  },
  "amoutonbrady@gmail.com": {
    "email": "amoutonbrady@gmail.com",
    "loc": 2,
    "percentageOfTotal": 0.0002448846654446922,
    "total_additions": 1,
    "total_deletions": 1,
  },
  "anton.yurovskykh@gmail.com": {
    "email": "anton.yurovskykh@gmail.com",
    "loc": 2,
    "percentageOfTotal": 0.0002448846654446922,
    "total_additions": 1,
    "total_deletions": 1,
  },
  "atif5801@gmail.com": {
    "email": "atif5801@gmail.com",
    "loc": 1,
    "percentageOfTotal": 0.0001224423327223461,
    "total_additions": 1,
    "total_deletions": 0,
  },
  "basaratali@gmail.com": {
    "email": "basaratali@gmail.com",
    "loc": 8,
    "percentageOfTotal": 0.0009795386617787688,
    "total_additions": 6,
    "total_deletions": 2,
  },
  "bryndyment@gmail.com": {
    "email": "bryndyment@gmail.com",
    "loc": 14,
    "percentageOfTotal": 0.0017141926581128453,
    "total_additions": 7,
    "total_deletions": 7,
  },
  "david@pilotinterative.ca": {
    "email": "david@pilotinterative.ca",
    "loc": 38,
    "percentageOfTotal": 0.004652808643449151,
    "total_additions": 19,
    "total_deletions": 19,
  },
  "enom@enom.ws": {
    "email": "enom@enom.ws",
    "loc": 2,
    "percentageOfTotal": 0.0002448846654446922,
    "total_additions": 1,
    "total_deletions": 1,
  },
  "github.100.fkeller@spamgourmet.com": {
    "email": "github.100.fkeller@spamgourmet.com",
    "loc": 109,
    "percentageOfTotal": 0.013346214266735724,
    "total_additions": 65,
    "total_deletions": 44,
  },
  "guido4000@users.noreply.github.com": {
    "email": "guido4000@users.noreply.github.com",
    "loc": 2,
    "percentageOfTotal": 0.0002448846654446922,
    "total_additions": 1,
    "total_deletions": 1,
  },
  "james@neodon.com": {
    "email": "james@neodon.com",
    "loc": 187,
    "percentageOfTotal": 0.02289671621907872,
    "total_additions": 179,
    "total_deletions": 8,
  },
  "joe@trusktr.io": {
    "email": "joe@trusktr.io",
    "loc": 4,
    "percentageOfTotal": 0.0004897693308893844,
    "total_additions": 1,
    "total_deletions": 3,
  },
  "modderme123@users.noreply.github.com": {
    "email": "modderme123@users.noreply.github.com",
    "loc": 2,
    "percentageOfTotal": 0.0002448846654446922,
    "total_additions": 1,
    "total_deletions": 1,
  },
  "nicolas.dermine@gmail.com": {
    "email": "nicolas.dermine@gmail.com",
    "loc": 2,
    "percentageOfTotal": 0.0002448846654446922,
    "total_additions": 1,
    "total_deletions": 1,
  },
  "praneybehl@gmail.com": {
    "email": "praneybehl@gmail.com",
    "loc": 3,
    "percentageOfTotal": 0.0003673269981670383,
    "total_additions": 2,
    "total_deletions": 1,
  },
  "rajeev22.18@gmail.com": {
    "email": "rajeev22.18@gmail.com",
    "loc": 10,
    "percentageOfTotal": 0.0012244233272234609,
    "total_additions": 5,
    "total_deletions": 5,
  },
  "ryansolid@gmail.com": {
    "email": "ryansolid@gmail.com",
    "loc": 816000,
    "percentageOfTotal": 99.91294350143441,
    "total_additions": 427412,
    "total_deletions": 388588,
  },
  "sewell.steve@gmail.com": {
    "email": "sewell.steve@gmail.com",
    "loc": 2,
    "percentageOfTotal": 0.0002448846654446922,
    "total_additions": 1,
    "total_deletions": 1,
  },
  "trusktr@gmail.com": {
    "email": "trusktr@gmail.com",
    "loc": 15,
    "percentageOfTotal": 0.0018366349908351914,
    "total_additions": 14,
    "total_deletions": 1,
  },
}
`;
